### PR TITLE
DataLoaderPrefetcher for quicker starts of training loops

### PR DESCRIPTION
**Pull request by pzelasko**
# What does this PR do ?

Removes/reduces the waiting on step 0 of a training loop for dataloading worker initialization. The waiting duration is specific to a given training setup and can range from a few seconds to a minute or two. An obvious factor is faster/slower disk/network, but also some settings e.g. using larger bucketing or shuffling buffers requires more time to pre-load data into memory. 

The main point is to trigger the actual initialization (which happens on `DataLoader.__iter__()`) at the time the dataloader is constructed, without blocking. Then it can pre-load the data concurrently to other stuff being initialized in training script and ideally it's ready to yield a mini-batch by the time the training loop starts.

**Collection**: ASR

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by pzelasko** [comment]
It causes a segmentation fault in GPU tests. Maybe I'll revisit this later but don't want to invest more time into it right now: closing.
